ui/app/AppLayouts/Wallet/services/dapps/types/SessionRequestWithAuth.qml

```diff
@@ -0,0 +1,57 @@
+import QtQuick 2.15
+
+import shared.stores 1.0
+
+SessionRequestResolved {
+    id: root
+
+    required property DAppsStore store
+
+    // Signal to execute the request. Emitted after successful authentication
+    // password and pin are the user's input for the authentication
+    signal execute(string password, string pin)
+    // Signal to reject the request. Emitted when the request is expired or rejected by the user
+    // hasError is true if the request was rejected due to an error
+    signal rejected(bool hasError)
+    // Signal when the authentication flow fails
+    signal authFailed()
+    
+    function accept() {
+        if (root.isExpired()) {
+            console.warn("Error: request expired")
+            root.reject(true)
+            return
+        }
+        storeConnections.enabled = true
+        store.authenticateUser(root.topic, root.requestId, root.accountAddress, "")
+    }
+
+    function reject(hasError) {
+        storeConnections.enabled = false
+        root.rejected(hasError)
+    }
+
+    Connections {
+        id: storeConnections
+        enabled: false
+        target: root.store
+        
+        function onUserAuthenticated(topic, id, password, pin, _) {
+            if (id == root.requestId && topic == root.topic) {
+                if (root.isExpired()) {
+                    console.warn("Error: request expired")
+                    root.reject(true)
+                    return
+                }
+                root.execute(password, pin)
+            }
+        }
+
+        function onUserAuthenticationFailed(topic, id) {
+            if (id === root.requestId && topic === root.topic) {
+                root.reject(true)
```

Comment:
Why do we need to send `reject` signal, isn't `authFailed` enough and the receiver will know how to handle that (do necessary actions)?

Reply:
Good point! It is!